modules/imgproc/src/spatialgradient.cpp

```diff
@@ -0,0 +1,276 @@
+    short* p_dy = dy.ptr<short>(0);
+
+    // Row, column indices
+    int i, j;
+
+    /* NOTE:
+     *
+     * Sobel-x: -1  0  1
+     *          -2  0  2
+     *          -1  0  1
+     *
+     * Sobel-y: -1 -2 -1
+     *           0  0  0
+     *           1  2  1
+     */
+
+    // No-SSE
+    int idx;
+
+    p_dx[0] = 0;   // Top-left corner
+    p_dy[0] = 0;
+    p_dx[W-1] = 0; // Top-right corner
+    p_dy[W-1] = 0;
+    p_dx[N-1] = 0; // Bottom-right corner
+    p_dy[N-1] = 0;
+    p_dx[N-W] = 0; // Bottom-left corner
+    p_dy[N-W] = 0;
+
+    // Handle special case: column matrix
+    if ( W == 1 )
+    {
+        for ( i = 1; i < H - 1; i++ )
+        {
+            p_dx[i] = 0;
+            p_dy[i] = 4*(p_src[i + 1] - p_src[i - 1]); // Should be 2?! 4 makes tests pass
+        }
+        return;
+    }
+
+    // Handle special case: row matrix
+    if ( H == 1 )
+    {
+        for ( j = 1; j < W - 1; j++ )
+        {
+            p_dx[j] = 4*(p_src[j + 1] - p_src[j - 1]); // Should be 2?! 4 makes tests pass
+            p_dy[j] = 0;
+        }
+        return;
+    }
+
+    // Do top row
+    for ( j = 1; j < W - 1; j++ )
+    {
+        idx = j;
+        p_dx[idx] = -(p_src[idx+W-1] + 2*p_src[idx-1] + p_src[idx+W-1]) +
+                     (p_src[idx+W+1] + 2*p_src[idx+1] + p_src[idx+W+1]);
+        p_dy[idx] = 0;
+    }
+
+    // Do right column
+    idx = 2*W - 1;
+    for ( i = 1; i < H - 1; i++ )
+    {
+        p_dx[idx] = 0;
+        p_dy[idx] = -(p_src[idx-W-1] + 2*p_src[idx-W] + p_src[idx-W-1]) +
+                     (p_src[idx+W-1] + 2*p_src[idx+W] + p_src[idx+W-1]);
+        idx += W;
+    }
+
+    // Do bottom row
+    idx = N - W + 1;
+    for ( j = 1; j < W - 1; j++ )
+    {
+        p_dx[idx] = -(p_src[idx-W-1] + 2*p_src[idx-1] + p_src[idx-W-1]) +
+                     (p_src[idx-W+1] + 2*p_src[idx+1] + p_src[idx-W+1]);
+        p_dy[idx] = 0;
+        idx++;
+    }
+
+    // Do left column
+    idx = W;
+    for ( i = 1; i < H - 1; i++ )
+    {
+        p_dx[idx] = 0;
+        p_dy[idx] = -(p_src[idx-W+1] + 2*p_src[idx-W] + p_src[idx-W+1]) +
+                     (p_src[idx+W+1] + 2*p_src[idx+W] + p_src[idx+W+1]);
+        idx += W;
+    }
+
+    // Do Inner area
+#if CV_SIMD128
+    // Characters in variable names have the following meanings:
+    // u: unsigned char
+    // s: signed int
+    //
+    // [row][column]
+    // m: offset -1
+    // n: offset  0
+    // p: offset  1
+    // Example: umn is offset -1 in row and offset 0 in column
+    v_uint8x16 v_um, v_un, v_up;
+    v_uint16x8 v_um1, v_um2, v_un1, v_un2, v_up1, v_up2;
+    v_int16x8 v_smm1, v_smm2, v_smn1, v_smn2, v_smp1, v_smp2,
+              v_snm1, v_snm2, v_snn1, v_snn2, v_snp1, v_snp2,
+              v_spm1, v_spm2, v_spn1, v_spn2, v_spp1, v_spp2,
+              v_two = v_setall_s16(2),
+              v_sdx1, v_sdx2, v_sdy1, v_sdy2;
+
+    for ( i = 1; i < H - 1; i++ )
+    {
+        // 16-column chunks at a time
```

Comment:
if you store pointers to (y-1)-th, y-th and y-th row in local variables and take into account the border extrapolation in the case of 0-th and (H-1)-th row, then the loop can be from 0 to H and you will not need separate code to process the top and bottom rows